components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +85,238 @@ public AbstractHTTPFunction() {
+                            Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.API, endpointURL)
+                            .resultMessage("Creating a connection for the external API call timed out.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                outcome = Constants.OUTCOME_TIMEOUT; // Timeout, retry if attempts left
                 LOG.error("Error while waiting to connect to " + endpointURL, e);
-                outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (SocketTimeoutException e) {
-                LOG.error("Error while waiting for data from " + endpointURL, e);
-                outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (IOException e) {
+            } else if (e instanceof IOException) {
+                outcome = Constants.OUTCOME_TIMEOUT; // Timeout, retry if attempts left
                 LOG.error("Error while calling endpoint. ", e);
-                outcome = Constants.OUTCOME_FAIL;
-            } catch (ParseException e) {
+            } else if (e instanceof ParseException) {
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.API, endpointURL)
+                            .resultMessage("Failed to parse the response from the external API.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                outcome = Constants.OUTCOME_FAIL; // Server error, retry if attempts left
                 LOG.error("Error while parsing response. ", e);
-                outcome = Constants.OUTCOME_FAIL;
+            } else {
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.API, endpointURL)
+                            .resultMessage("Received unknown exception from external API call.")
```

Comment:
Shall we say "Recieved an error while invoking the external API". This is because it's not the server that returned the error which seems to be the idea if we use the phrase "from external API"

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/dd0c0830ab854ed1533c10099ec4bb705ddeb218